docs/GDPs/gdp-0002-regions.rst

```diff
@@ -0,0 +1,275 @@
+    def main_ub(uc: Field, vc: Field, cosa: Field, rsina: Field, ub: Field, dt5: float):​
+        with computation(PARALLEL), interval(...):​
+            ub = dt5 * (uc[0, -1, 0] + uc - ​(vc[-1, 0, 0] + vc) * cosa) * rsina​
+
+    @gtscript.stencil()​
+    def x_edge_ub(ut: Field, ub: Field, dt5: float):
+        with computation(PARALLEL), interval(...):​
+            ub = dt5 * (ut[0, -1, 0] + ut)​
+
+    @gtscript.stencil()​
+    def y_edge_ub(ut: Field, ub: Field, *, dt4: float):
+        with computation(PARALLEL), interval(...):
+            ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
+
+    # Requires a python function to call stencils on each region
+    def ubke(uc, vc, ut, ub, dt5, dt4, grid):​
+        domain_y_edge = (grid.ni, 1, grid.nz)
+        domain_x_edge = (1, grid.nj, grid.nz)
+        main_ub(uc, vc, grid.cosa, grid.rsina, ub, dt5=dt5, ​
+                origin=(grid.local_istart, grid.local_jstart, 0),
+                domain=(grid.ni, grid.nj, grid.nz))​
+        if grid.west_edge:​
+            x_edge_ub(ut, ub, dt5=dt5, ​origin=(grid.local_istart, grid.local_jstart, 0), ​domain=domain_x_edge)​
+        if grid.south_edge:
+            y_edge(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jstart, 0), domain=domain_y_edge)
+        if grid.north_edge:
+            y_edge_ub(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jend, 0), domain=domain_y_edge)
+        if grid.east_edge:
+            x_edge_ub(ut, ub, dt5=dt5, origin=(grid.local_iend, grid.local_jstart, 0), domain=domain_x_edge)
+
+The specific feature that we are proposing is the addition of a ``with horizontal()`` context that specifies the horizontal iteration space (over parallel axes) using ``region`` objects.
+The ``region`` object's ``__getitem__`` method uses `splitters` to select a region of the stencil computational domain in the parallel ``I`` and ``J`` axes.
+
+Using this specification, the example is transformed into:
+
+.. code-block:: python
+
+    istart = I[0] if grid.west_edge else None
+    jstart = J[0] if grid.south_edge else None
+    iend = I[-1] if grid.east_edge else None
+    jend = J[-1] if grid.north_edge else None
+
+    @gtscript.stencil()
+    def stencil(uc: Field, vc: Field, cosa: Field, rsina: Field, ut: Field, ub: Field, dt5: float, dt4: float):
+        from __externals__ import istart, iend, jstart, jend
+        with computation(PARALLEL), interval(...):
+            ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
+            with horizontal(region[istart, :], region[iend, :]):
+                ub = dt5 * (ut[0, -1, 0] + ut)
+            with horizontal(region[:, jstart], region[:, jend]):
+                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
+
+Notice that distributed domain decomposition works with this feature - if the external has the value ``None``, then any ``region`` objects using this in its slices ignores it.
```

Comment:
Yes, that's it! Note that this is not quite watertight, but good enough for now. We can have a situation of a 3x1 processor layout, then on the middle processor this region will fail to execute: 
```
region[start : stop, :]
```
since `start` and `stop` are both `None`. Do you have a suggestion to get around that?